### ISSUE TITLE
Common subexpression elimination (cse)

### ISSUE THREAD

**Issue by 363734** [CPMpy/cpmpy#360]
Following last week discussion. We notice there is no cse implemented. And I checked the decompositions of a few global expressions, they are duplicating (a lot) their arguments. Which is not an issue if the arg is a simple var, but may become very problematic if the argument is any more complicated expressions as it will duplicate constraints posted within the solvers. 

I’m attending a talk of Guido about minizinc where he just recall the importance of if.

I think this should be added to the next release if possible.

**Comment by Wout4** [comment]
I think we all agree this has to be added at some point, I'd be interested to work on it. We also have it as a master thesis topic for next year, so we'll see if someone picks it up

**Comment by IgnaceBleukx** [comment]
I fully agree we should implement this at some point, but this does not have priority to get into the next release for me.
A first (very naive) version could be to cache calls to `get_or_make_var`. This will at least ensure `flatten_constraint` does not create variables which are identically defined  by constraints.

**Comment by 363734** [comment]
Yeah, I didn't meant having the most optimized one for the next release, but a basic naive one would be nice already.

**Comment by tias** [comment]
Yes, `get_or_make_var()` is the obvious one, plus in flatten checking the flat `==` expressions for equivalent expressions to add to the CSE dict.

Bonus would be for `decompose_in_tree()` to also check GlobalFunction reuses.

There might be a bit of performance penalty in flatten because of the (recursive) string-based way we implemented `__hash__()`, not sure.
